src/treelearner/feature_histogram.hpp

```diff
@@ -211,13 +213,28 @@ class FeatureHistogram {
 
 private:
   /*!
+  * \brief Return the L1 regularized sum_gradients
+  * \param sum_gradients
+  * \return regularized sum_gradients
+  */
+  double L1RegularizedGradients(double sum_gradients) const {
+  if (sum_gradients > lambda_l1_) return sum_gradients - lambda_l1_;
+  if (sum_gradients < -lambda_l1_) return sum_gradients + lambda_l1_;
+  return 0.0f;
+  }
+
+  /*!
   * \brief Calculate the split gain based on sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (lambda_l1_ != 0.0f) {
```

Comment:
Yes, that's correct. The condition just prevents an unnecessary function call (and one or two additional `if`'s) when `lambda_l1_ == 0.0f`. The same goes for the`if` in `CalculateSplittedLeafOutput`. If you don't think this will make much (any?) difference in terms of performance, we should just get rid of this.